File: 30% Group Project/Group B2/docker-socket-project/c-server2/server2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <mysql/mysql.h>
#include <time.h>

#define PORT 8084
#define DB_HOST "database"
#define DB_USER "root"
#define DB_PASS "rootpassword"
#define DB_NAME "socket_db"

void update_database_points(const char* username) {
    MYSQL *conn = mysql_init(NULL);
    
    if (!mysql_real_connect(conn, DB_HOST, DB_USER, DB_PASS, DB_NAME, 0, NULL, 0)) {
        fprintf(stderr, "Database connection failed: %s\n", mysql_error(conn));
        mysql_close(conn);
        return;
    }
    
    char query[256];
    snprintf(query, sizeof(query),
             "UPDATE user_points SET points = points + 1, datetime_stamp = NOW() WHERE user = '%s'",
             username);
    
    if (mysql_query(conn, query)) {
        fprintf(stderr, "Update failed: %s\n", mysql_error(conn));
    } else {
        printf("Updated points for %s\n", username);
    }
    
    mysql_close(conn);
}

void get_latest_points(const char* username, char* buffer, size_t buffer_size) {
    MYSQL *conn = mysql_init(NULL);
    
    if (!mysql_real_connect(conn, DB_HOST, DB_USER, DB_PASS, DB_NAME, 0, NULL, 0)) {
        snprintf(buffer, buffer_size, "Database connection failed");
        return;
    }
    
    char query[256];
    snprintf(query, sizeof(query),
             "SELECT points, datetime_stamp FROM user_points WHERE user = '%s'",
             username);
    
    if (mysql_query(conn, query) == 0) {
        MYSQL_RES *result = mysql_store_result(conn);
        if (result) {
            MYSQL_ROW row = mysql_fetch_row(result);
            if (row) {
                snprintf(buffer, buffer_size, "User: %s, Points: %s, Last Updated: %s", 
                         username, row[0], row[1]);
            }
            mysql_free_result(result);
        }
    }
    
    mysql_close(conn);
}

int main() {
    int server_fd, new_socket;
    struct sockaddr_in address;
    int opt = 1;
    int addrlen = sizeof(address);
    char buffer[1024] = {0};
    
    // Create socket
    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
        perror("socket failed");
        exit(EXIT_FAILURE);
    }
    
    // Set socket options
    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt))) {
        perror("setsockopt");
        exit(EXIT_FAILURE);
    }
    
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY;
    address.sin_port = htons(PORT);
    
    // Bind socket
    if (bind(server_fd, (struct sockaddr *)&address, sizeof(address)) < 0) {
        perror("bind failed");
        exit(EXIT_FAILURE);
    }
    
    // Listen
    if (listen(server_fd, 3) < 0) {
        perror("listen");
        exit(EXIT_FAILURE);
    }
    
    printf("C Server listening on port %d\n", PORT);
    
    // Auto-update points every 30 seconds
    pid_t pid = fork();
    if (pid == 0) {
        // Child process for periodic updates
        while(1) {
            sleep(30);
            update_database_points("c_user2");
        }
    }
    
    // Main process for handling client connections
    while(1) {
        if ((new_socket = accept(server_fd, (struct sockaddr *)&address, (socklen_t*)&addrlen)) < 0) {
            perror("accept");
            continue;
        }
        
        // Read client request
        read(new_socket, buffer, 1024);
        
        if (strstr(buffer, "GET_POINTS") != NULL) {
            char response[1024];
            get_latest_points("c_user2", response, sizeof(response));
            send(new_socket, response, strlen(response), 0);
        }
        
        close(new_socket);
    }
    
    return 0;
}

File: 30% Group Project/Group B2/docker-socket-project/python-server2/server2.py
import socket
import time
import threading
import mysql.connector
from datetime import datetime

# Database configuration
DB_CONFIG = {
    'host': 'database',
    'user': 'root',
    'password': 'rootpassword',
    'database': 'socket_db'
}

PORT = 8085
USERNAME = "python_user2"

def update_database_points():
    """Update points for python_user2 every 30 seconds"""
    while True:
        try:
            conn = mysql.connector.connect(**DB_CONFIG)
            cursor = conn.cursor()
            
            query = """
            UPDATE user_points 
            SET points = points + 1, datetime_stamp = NOW() 
            WHERE user = %s
            """
            
            cursor.execute(query, (USERNAME,))
            conn.commit()
            print(f"Updated points for {USERNAME} at {datetime.now()}")
            
            cursor.close()
            conn.close()
        except Exception as e:
            print(f"Database update error: {e}")
        
        time.sleep(30)

def get_latest_points():
    """Get latest points from database"""
    try:
        conn = mysql.connector.connect(**DB_CONFIG)
        cursor = conn.cursor()
        
        query = """
        SELECT points, datetime_stamp 
        FROM user_points 
        WHERE user = %s
        """
        
        cursor.execute(query, (USERNAME,))
        result = cursor.fetchone()
        
        cursor.close()
        conn.close()
        
        if result:
            return f"User: {USERNAME}, Points: {result[0]}, Last Updated: {result[1]}"
        else:
            return f"No data found for {USERNAME}"
            
    except Exception as e:
        return f"Error: {str(e)}"

def handle_client(client_socket):
    """Handle client requests"""
    try:
        request = client_socket.recv(1024).decode('utf-8')
        
        if "GET_POINTS" in request:
            response = get_latest_points()
            client_socket.send(response.encode('utf-8'))
        else:
            client_socket.send(b"Invalid request")
            
    except Exception as e:
        print(f"Client handling error: {e}")
    finally:
        client_socket.close()

def start_server():
    """Start the socket server"""
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind(('0.0.0.0', PORT))
    server.listen(5)
    
    print(f"Python Server listening on port {PORT}")
    
    # Start periodic update thread
    update_thread = threading.Thread(target=update_database_points, daemon=True)
    update_thread.start()
    
    # Accept client connections
    while True:
        client_socket, addr = server.accept()
        print(f"Connection from {addr}")
        
        client_thread = threading.Thread(target=handle_client, args=(client_socket,))
        client_thread.start()

if __name__ == "__main__":
    start_server()
